### ISSUE TITLE
Video RAG example

### ISSUE THREAD

**Issue by ardge-timwu** [ardge-timwu/rag-experiments#45]
https://www.llamaindex.ai/blog/multimodal-rag-for-advanced-video-processing-with-llamaindex-lancedb-33be4804822e

need some adjustments
* lancedb -- as is
* GPT-4V -> LLava
* llamaindex -- as is

## code
* https://github.com/run-llama/llama_index/blob/main/docs/docs/examples/multi_modal/multi_modal_video_RAG.ipynb

## related references
* https://developer.nvidia.com/blog/an-easy-introduction-to-multimodal-retrieval-augmented-generation/
* https://www.youtube.com/watch?v=LF7I6raAIL4
* https://arxiv.org/abs/2402.15116

## related code
https://github.com/run-llama/llama_index/tree/main/docs/docs/examples/multi_modal
* [llava_multi_modal_tesla_10q.ipynb](https://github.com/run-llama/llama_index/blob/main/docs/docs/examples/multi_modal/llava_multi_modal_tesla_10q.ipynb)
* [gpt4v_multi_modal_retrieval.ipynb](https://github.com/run-llama/llama_index/blob/main/docs/docs/examples/multi_modal/gpt4v_multi_modal_retrieval.ipynb)

## testbed
http://172.16.1.82:8888/lab/tree/llama_index/docs/docs/examples/multi_modal/multi_modal_video_RAG.ipynb

**Comment by ardge-timwu** [comment]
**Originally posted by @timwu-ipevo** *(2024/7/4)*

Key questions about the approach
<details>
<summary>About the example code given in section Building the Multi-Modal Index and Vector Store, I wonder how `MultiModalVectorStoreIndex` works.  Since the input document are actually images, how does llamaindex know how to vectorize an image?</summary>
<i>
The `MultiModalVectorStoreIndex` in the example code leverages OpenAI’s CLIP model to generate embeddings for both textual and visual data. Here's how it works:

1. **Image Vectorization**: The `MultiModalVectorStoreIndex` uses the CLIP model to convert images into vector embeddings. CLIP (Contrastive Language–Image Pretraining) is designed to understand images and their corresponding textual descriptions. It can generate embeddings for images by processing them through its neural network, which has been trained on a large dataset of images and associated text.

2. **Text Vectorization**: Similarly, the CLIP model can generate embeddings for text. The text data extracted from the video (e.g., transcriptions) is processed to create vector embeddings.

3. **Storage in LanceDB**: These embeddings (both for images and text) are stored and managed in LanceDB, a vector database. The `LanceDBVectorStore` class is used to handle the storage and retrieval of these embeddings.

Here’s a simplified breakdown of the process:

- **Image Embeddings**: When the images are processed, the `MultiModalVectorStoreIndex` uses the CLIP model to generate vector embeddings for each image. These embeddings capture the semantic content of the images.

- **Text Embeddings**: The text data is similarly processed to generate vector embeddings using the CLIP model.

- **Index Construction**: The generated embeddings are then stored in LanceDB, creating a multimodal index that can be queried for relevant information.

The code snippet provided in the document shows how to set up this process:

```python
from llama_index.indices.multi_modal.base import MultiModalVectorStoreIndex
from llama_index import SimpleDirectoryReader, StorageContext
from llama_index.vector_stores import LanceDBVectorStore

text_store = LanceDBVectorStore(uri="lancedb", table_name="text_collection")
image_store = LanceDBVectorStore(uri="lancedb", table_name="image_collection")

storage_context = StorageContext.from_defaults(vector_store=text_store, image_store=image_store)
documents = SimpleDirectoryReader(output_folder).load_data()

index = MultiModalVectorStoreIndex.from_documents(documents, storage_context=storage_context)
```

In this code:
- `SimpleDirectoryReader` loads the data (images and text) from the specified folder.
- `LanceDBVectorStore` instances (`text_store` and `image_store`) are created to manage the storage of text and image embeddings.
- `StorageContext` combines these vector stores.
- `MultiModalVectorStoreIndex.from_documents` processes the documents to generate and store the embeddings in the vector stores.

By using the CLIP model, `llamaindex` can effectively vectorize images and text, enabling the creation of a robust multimodal index for efficient retrieval and analysis.
</i>
</details>

<details>
<summary>building index(embeddings) for every frame in the video with CLIP sounds a time consuming task.  Is there any argument about this approach?</summary>
<i>
Yes, building embeddings for every frame in a video using CLIP can indeed be a time-consuming task, especially for long videos or high frame rates. Here are some considerations and potential arguments regarding this approach:

### Arguments in Favor

1. **Rich Contextual Information**: By generating embeddings for every frame, you capture a detailed and comprehensive representation of the video's content. This can be particularly useful for applications requiring high precision, such as security surveillance or detailed educational content analysis.

2. **Improved Retrieval Accuracy**: More embeddings mean more data points for the retrieval system to work with, potentially improving the accuracy and relevance of the retrieved information.

3. **Flexibility in Query Responses**: Having embeddings for every frame allows the system to respond to a wide variety of queries with high specificity, as it can pull from a rich dataset of visual and textual information.

### Arguments Against

1. **Time and Resource Intensive**: Processing every frame can be extremely time-consuming and computationally expensive. This might not be feasible for real-time applications or for processing large volumes of video data.

2. **Redundancy**: Consecutive frames in a video often contain very similar information. Generating embeddings for every frame might lead to a lot of redundant data, which could be inefficient in terms of storage and processing.

3. **Diminishing Returns**: The additional value gained from processing every single frame might not justify the extra computational cost. In many cases, processing fewer frames (e.g., one frame per second) might provide sufficient information for most applications.

### Potential Optimizations

1. **Frame Sampling**: Instead of processing every frame, you can sample frames at regular intervals (e.g., one frame every second or every few seconds). This reduces the number of frames to process while still capturing the video's overall content.

2. **Key Frame Extraction**: Use algorithms to detect and process only key frames, which are frames that represent significant changes or important scenes in the video. This approach can reduce redundancy and focus on the most informative parts of the video.

3. **Parallel Processing**: Utilize parallel processing techniques and hardware accelerators (like GPUs) to speed up the embedding generation process. This can help mitigate the time cost associated with processing large numbers of frames.

4. **Dynamic Adjustment**: Implement a dynamic approach where the frame sampling rate can be adjusted based on the video's content complexity. For example, scenes with a lot of motion or changes might require more frequent sampling compared to static scenes.

### Conclusion

While generating embeddings for every frame using CLIP provides a detailed and rich dataset, it is often not practical due to the high computational cost. Optimizations like frame sampling, key frame extraction, and parallel processing can help balance the trade-off between accuracy and efficiency. The choice of approach should be guided by the specific requirements and constraints of the application in question.

</i>
</details>
<details><summary>
according to section 2 `Video Processing`, the fps 0.2 is quite low.  The overhead I said is probably not an issue.  I just wonder how does the system align embeddings extracted from audio text and video frames.  Two embeddings extracted at relatively closed timing might be related or unrelated.
</summary>
<i>
The alignment of embeddings extracted from audio text and video frames is indeed a crucial aspect of creating a coherent multimodal index. Here’s how the system can handle this alignment:

### Temporal Alignment Strategy

1. **Timestamp Synchronization**: Both video frames and audio text can be associated with timestamps. By aligning the embeddings based on these timestamps, the system can ensure that the visual and textual data are temporally synchronized.

2. **Chunk-Based Processing**: Instead of processing individual frames and audio snippets independently, the system can divide the video into chunks (e.g., 5-second intervals). Each chunk would contain a set of frames and the corresponding audio text. Embeddings are then generated for these chunks, ensuring that the visual and textual data within each chunk are related.

### Detailed Steps

1. **Extract Frames and Audio**: As described in the document, frames are extracted at a rate of 0.2 fps (one frame every 5 seconds). The audio is also extracted and transcribed into text.

2. **Generate Embeddings**:
   - **Frames**: For each extracted frame, generate an embedding using the CLIP model.
   - **Audio Text**: For the audio text corresponding to each 5-second chunk, generate an embedding using the CLIP model.

3. **Align Embeddings**:
   - **Temporal Matching**: Since both the frames and the audio text are extracted in 5-second intervals, they can be aligned based on these intervals. For example, the embedding for the frame at the 5-second mark is aligned with the embedding for the audio text transcribed from 0 to 5 seconds.
   - **Chunk Embeddings**: Create a combined embedding for each chunk by concatenating or averaging the embeddings of the frames and the audio text within that chunk.

### Example Code for Alignment

Here’s a simplified example of how you might align the embeddings:

```python
import numpy as np
from moviepy.editor import VideoFileClip
import speech_recognition as sr
from transformers import CLIPProcessor, CLIPModel

# Initialize CLIP model and processor
model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")

# Function to generate embeddings for frames
def generate_frame_embeddings(frames):
    inputs = processor(images=frames, return_tensors="pt", padding=True)
    embeddings = model.get_image_features(**inputs)
    return embeddings

# Function to generate embeddings for text
def generate_text_embeddings(texts):
    inputs = processor(text=texts, return_tensors="pt", padding=True)
    embeddings = model.get_text_features(**inputs)
    return embeddings

# Extract frames and audio text
video_path = "./video_data/input_vid.mp4"
clip = VideoFileClip(video_path)
frames = [clip.get_frame(t) for t in np.arange(0, clip.duration, 5)]
audio_path = "./mixed_data/output_audio.wav"
recognizer = sr.Recognizer()
audio = sr.AudioFile(audio_path)
with audio as source:
    audio_data = recognizer.record(source)
    text_data = recognizer.recognize_whisper(audio_data)

# Split text data into chunks (assuming text_data is a single string)
text_chunks = [text_data[i:i+5] for i in range(0, len(text_data), 5)]

# Generate embeddings
frame_embeddings = generate_frame_embeddings(frames)
text_embeddings = generate_text_embeddings(text_chunks)

# Align embeddings by chunk
aligned_embeddings = []
for frame_emb, text_emb in zip(frame_embeddings, text_embeddings):
    combined_emb = np.concatenate((frame_emb, text_emb), axis=0)
    aligned_embeddings.append(combined_emb)

# Store or process aligned embeddings as needed
```

### Handling Misalignment

- **Contextual Overlap**: Ensure that the text chunks and frames cover overlapping time periods to handle slight misalignments.
- **Post-Processing**: Use additional post-processing techniques to refine the alignment, such as checking for semantic similarity between adjacent chunks.

### Conclusion

By synchronizing frames and audio text based on timestamps and processing them in chunks, the system can effectively align the embeddings. This ensures that the visual and textual data are contextually related, enabling more accurate and coherent multimodal indexing and retrieval.

</i></details>

**Comment by ardge-timwu** [comment]
**Originally posted by @timwu-ipevo** *(2024/7/5)*

### experiment ollama+llava inference

It can perform inference on multiple images, but the capacity is limited. Two images are typically manageable; attempting more than two often causes Ollama to hang or become unresponsive.
```py 
plist=['0014', '0050']#, '0087']#, '0144']
imgfns = [f'mixed_data/frame{pno}.png' for pno in plist] 
print(imgfns)
imgbss = [ read_png_to_bytes(imgfn) for imgfn in imgfns]
response = oclient.chat(model='llava', messages=[
  {
    'role': 'user',
    'content': 'Describe these images as detail as possible',
    'images': imgbss
  },
])
response
```
```
 七  05 09:29:03 ubpc ollama[1821408]: [GIN] 2024/07/05 - 09:29:03 | 200 |  6.641740187s |     172.16.1.82 | POST     "/api/chat"
 七  05 09:29:28 ubpc ollama[1821408]: [GIN] 2024/07/05 - 09:29:28 | 200 |  7.860082708s |     172.16.1.82 | POST     "/api/chat"
```

**Comment by ardge-timwu** [comment]
**Originally posted by @timwu-ipevo** *(2024/7/5)*

The current challenge lies in the final phase of RAG, which involves LLM inference. While this may only require a single round of chat completion API, it often involves a substantial amount of context. This scenario demands that the LLM possess an adequately long context capacity. For instance, with a typical retrieval result of 5 images, the context easily surpasses the capacity of a local MLLM such as LLaVA.

**Comment by ardge-timwu** [comment]
**Originally posted by @timwu-ipevo** *(2024/7/5)*

llava's context length is 32K and embedding length is 4K.
```
$ollama show llava
  Model                             
        arch                    llama                
        parameters              7B                   
        quantization            Q4_0                 
        context length          32768                
        embedding length        4096                 
                                          
  Projector                         
        arch                            clip      
        parameters                      311.89M   
        projector type                  mlp       
        embedding length                1024      
        projection dimensionality       768       
                                          
  Parameters                        
        stop    "[INST]"                             
        stop    "[/INST]"                            
                                          
  License                           
        Apache License                           
        Version 2.0, January 2004  
```

just for reference
```
ollama show changtimwu/llama3-8b-twllm
  Model                   
        arch                    llama     
        parameters              8.0B      
        quantization            Q4_K_M    
        context length          8192      
        embedding length        4096      
```
